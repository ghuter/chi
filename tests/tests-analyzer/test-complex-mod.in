f :: fun () i32 {
	x :: 10;
	HasherI32.hash(@x);
	return HasherI32.hash(@x);
}

Hasher :: signature [T] {
	AStruct :: struct {
		x: ^T;
	}
	hash :: fun (v: ^T) i32;
}

HasherI32 :: impl Hasher [i32] {
	hash :: fun (v: ^i32) i32 {
		i :: 10;
		a :: AStruct {x: @i};
		return ^a.x;
	}
}

HasherI8 :: impl Hasher [i8] {
	hash :: fun (v: ^i8) i32 {
		return 0;
	}
}

HashMap :: signature [K, V] (H: Hasher[K]) {
	// Generic type must be used as pointer because they are translated into void*.
	Node :: struct {
		key: ^K;
		value: ^V;
	}

	Map :: struct {
		nodes: ^Node;
		nnode: u32;
	}

	insert :: fun (map: Map, key: ^K, value: ^V) bool;
	get :: fun (map: Map, key: ^K) ^V;
}

MachinChose :: impl HashMap [i32, i32] (H: HasherI32) {
	insert :: fun (map: Map, key: ^i32, value: ^i32) bool {
		return true;
	}

	get :: fun (map: Map, key: ^i32) ^i32 {
		x :: 10;
		return @x;
	}
}

SHashMap :: skeleton HashMap {

	insert :: fun (map: Map, key: ^K, value: ^V) bool {
		return true;
	}

	get :: fun (map: Map, key: ^K) ^V {
		nodes :: map.nodes;
		n :: ^nodes;

		H.hash(key);
		x :: H.hash(key);
		return n.value;
	}
}

I32Map :: define SHashMap [i32, i32] (H: HasherI32);
