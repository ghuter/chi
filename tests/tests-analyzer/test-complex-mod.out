SFUN(f, {SSEQ(const SDECL(x, ECSTI(10)) : i32, SSEQ(SACCESSMOD(HasherI32, SCALL(hash, EUNOP(AT, EMEM(x) : i32) : ^i32)), SRETURN(ECALL(EACCESS(EMEM(HasherI32) : ?.hash) : ?, EUNOP(AT, EMEM(x) : i32) : ^i32) : i32)))}) : i32
SMODSIGN(Hasher, [T], (), {SSTRUCT(AStruct, x : ^T), SSIGN(hash, v : ^T) : i32})
SMODSIGN(HashMap, [K, V], (H: Hasher[K]), {SSTRUCT(Node, key : ^K, value : ^V), SSTRUCT(Map, nodes : ^Node, nnode : u32), SSIGN(insert, map : Map, key : ^K, value : ^V) : bool, SSIGN(get, map : Map, key : ^K) : ^V})
SMODIMPL(HasherI32, [i32], (), | T => i32 |, {SFUN(hash, v : ^i32, {SSEQ(const SDECL(i, ECSTI(10)) : i32, SSEQ(const SDECL(a, ESTRUCT(AStruct { <x : EUNOP(AT, EMEM(i) : i32) : ^i32> })) : AStruct, SRETURN(EUNOP(DEREF, EACCESS(EMEM(a) : AStruct.x) : ^i32) : i32)))}) : i32}) : Hasher
SMODIMPL(HasherI8, [i8], (), | T => i8 |, {SFUN(hash, v : ^i8, {SRETURN(ECSTI(0))}) : i32}) : Hasher
SMODIMPL(MachinChose, [i32, i32], (H: HasherI32[]), | K => i32 | V => i32 |, {SFUN(insert, map : Map, key : ^i32, value : ^i32, {SRETURN(ETRUE)}) : bool, SFUN(get, map : Map, key : ^i32, {SSEQ(const SDECL(x, ECSTI(10)) : i32, SRETURN(EUNOP(AT, EMEM(x) : i32) : ^i32))}) : ^i32}) : HashMap
SMODSKEL(SHashMap, SFUN(insert, map : Map, key : ^K, value : ^V, {SRETURN(ETRUE)}) : bool, SFUN(get, map : Map, key : ^K, {SSEQ(const SDECL(nodes, EACCESS(EMEM(map) : Map.nodes) : ^Node) : ^Node, SSEQ(const SDECL(n, EUNOP(DEREF, EMEM(nodes) : ^Node) : Node) : Node, SSEQ(SACCESSMOD(H, SCALL(hash, EMEM(key) : ^K)), SSEQ(const SDECL(x, ECALL(EACCESS(EMEM(H) : ?.hash) : ?, EMEM(key) : ^K) : i32) : i32, SRETURN(EACCESS(EMEM(n) : Node.value) : ^V)))))}) : ^V}) : HashMap
SMODDEF(I32Map, [i32, i32], (H: HasherI32[]), | K => i32 | V => i32 |) : SHashMap
